modules/gapi/samples/pipeline_modeling_tool/pipeline.hpp

```diff
@@ -71,50 +76,88 @@ class Pipeline {
 
 protected:
     virtual void    _compile() = 0;
-    virtual int64_t run_iter() = 0;
+    virtual double  run_iter() = 0;
     virtual void    init() {};
     virtual void    deinit() {};
 
+    void prepareOutputs();
+
     std::string                  m_name;
     cv::GComputation             m_comp;
     std::shared_ptr<DummySource> m_src;
     StopCriterion::Ptr           m_stop_criterion;
     cv::GCompileArgs             m_args;
     size_t                       m_num_outputs;
     PerfReport                   m_perf;
+
+    cv::GRunArgsP                m_pipeline_outputs;
+    std::vector<cv::Mat>         m_out_mats;
+    int64_t                      m_start_ts;
+    double                       m_latency;
 };
 
 Pipeline::Pipeline(std::string&&                  name,
                    cv::GComputation&&             comp,
                    std::shared_ptr<DummySource>&& src,
                    StopCriterion::Ptr             stop_criterion,
                    cv::GCompileArgs&&             args,
-                   const size_t                   num_outputs)
+                   const size_t                   num_outputs,
+                   const double                   latency)
     : m_name(std::move(name)),
       m_comp(std::move(comp)),
       m_src(std::move(src)),
       m_stop_criterion(std::move(stop_criterion)),
       m_args(std::move(args)),
-      m_num_outputs(num_outputs) {
+      m_num_outputs(num_outputs),
+      m_latency(latency) {
     m_perf.name = m_name;
 }
 
 void Pipeline::compile() {
     m_perf.warmup_time =
-        utils::measure<std::chrono::milliseconds>([this]() {
+        utils::measure<utils::double_ms_t>([this]() {
         _compile();
     });
 }
 
+void Pipeline::prepareOutputs() {
+    // NB: N-1 buffers + timestamp.
+    m_out_mats.resize(m_num_outputs - 1);
+    for (auto& m : m_out_mats) {
+        m_pipeline_outputs += cv::gout(m);
+    }
+    m_pipeline_outputs += cv::gout(m_start_ts);
+}
+
 void Pipeline::run() {
     using namespace std::chrono;
 
+    // NB: Allocate outputs for execution
+    prepareOutputs();
+
+    // NB: Warm-up iteration invalidates source state
+    // so need to copy it
+    auto orig_src = m_src;
+    auto copy_src = std::make_shared<DummySource>(*m_src);
```

Comment:
Do you need `std::make_shared` here?

Reply:
I guess so, how it can be done without make_shared?

Reply:
Got it, thanks